libbeat/outputs/backoff.go

```diff
@@ -0,0 +1,51 @@
+package outputs
+
+import (
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/publisher"
+)
+
+type backoffClient struct {
+	client NetworkClient
+
+	done    chan struct{}
+	backoff *common.Backoff
+}
+
+// WithBackoff wraps a NetworkClient, adding exponential backoff support to a network client if connection/publishing failed.
+func WithBackoff(client NetworkClient, init, max time.Duration) NetworkClient {
```

Comment:
yeah. It's kind of a wrapper for NetworkClient. This will be moved closer to the libbeat pipeline itself. Backoff strategy + parameters will become config parameters in `outputs.Group`.